templating/base_render_options.go

```diff
@@ -152,6 +159,54 @@ func LoadAndRenderSteps(template *Template, opts *BaseRenderOptions) (string, er
 	return rendered, nil
 }
 
+// ResolveAndMergeSecrets parses the secrets in the template, resolves them using vault providers and adds them to the values map.
+func resolveAndMergeSecrets(ctx context.Context, template *Template, secretResolver *SecretResolver, valueMap *Values) error {
+
+	// Cheap optimization to skip the secrets merging if it doesn't contain "secrets" string in it. Note that the task can
+	// have the string secrets but may not essentially the secrets section.
+	if !strings.Contains(string(template.Data), "secrets") {
+		return nil
+	}
+
+	// At first render the template with existing values to render templatized vaules for secrets.
+	values := *valueMap
+	values["Secrets"] = Values{}
+	engine := NewEngine()
+	rendered, err := engine.Render(template, values)
+	if err != nil {
+		return errors.Wrap(err, "failed to render the template")
+	}
+
+	if rendered == "" {
+		return errors.New("rendered template was empty")
+	}
+
+	// Unmarshall the template to Task and get all secrets defined in the template.
+	task, err := graph.NewTaskFromString(rendered)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse template to create task")
+	}
+
+	// If no secrets found return.
+	if len(task.Secrets) == 0 {
+		return nil
+	}
+
+	if secretResolver == nil {
+		return errors.New("secret resolver is required")
+	}
+
+	resolvedSecrets, err := secretResolver.ResolveSecrets(ctx, task.Secrets)
```

Comment:
seems like the instantiation of the secret resolver should go before this line so that it's not unnecessarily allocated

Reply:
If we instantiate the secretresolver over here, there are two concerns. 1. We need to pass all initialization parameters for secret resolver to this function 2. If we hard code the secret resolver, it will be hard to unit test this function by mocking the resolver